### PR TITLE
Dockerfile for standalone cherami-server

### PR DESCRIPTION
**Pull request by thuningxu**
This PR contains files to build and run a docker container to run Cherami in single-server, standalone mode. The container includes Cassandra and all the Cherami binaries. The image is currently public at Docker registry as `thuningxu/cherami-server:latest`.

To start the container with name "cherami":
`$ docker run -d --name cherami thuningxu/cherami-server-standalone:latest`

Please wait about 30 seconds for the Cassandra and Cherami to start up. Then, you can connect to it using the container IP address:

`$ docker exec cherami hostname --ip-address`
`172.17.0.2`

You can use the tools inside the container to interact with Cherami, for example, create a destination:
`$ docker exec cherami ./cherami-cli --env=prod --hostport=172.17.0.2:4922 c d "/test/cherami"`

Create a consumer group:
`$ docker exec cherami ./cherami-cli --env=prod --hostport=172.17.0.2:4922 c c "/test/cherami" "/test/cherami_reader"`

Publish from terminal (need `-it` flag for docker):
`$ docker exec -it cherami ./cherami-cli --env=prod --hostport=172.17.0.2:4922 p "/test/cherami"`

Consume:
`$ docker exec cherami ./cherami-cli --env=prod --hostport=172.17.0.2:4922 r "/test/cherami" "/test/cherami_reader"`

### CODE REVIEW

File: docker/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# start up cassandra
+pushd /
+./docker-entrypoint.sh cassandra
+popd
+sleep 20
+
+export HOST_IP=`hostname --ip-address`
+
+# setup schema
+sed -i.bak "s/'replication_factor':\ '5'/'replication_factor':\ '1'/g" clients/metadata/schema/metadata_keyspace.cql
```
Comment (aravindvs) [inline docker/entrypoint.sh]: Can the RF be a config option?

**Comment by aravindvs** [comment]
This is great! I am wondering if we need to expose the "RingHosts" config as part of the template.yaml as well? That way we can use the same mechanism to form a cherami cluster right? i.e, modify the RingHosts to include the ip address of all your machines (or a subset of your machines that are considered a seed) where you will be running cherami and start the docker container on all machines?

File: docker/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ln -sf `which cqlsh` /usr/local/bin/cqlsh
+CQLSH_HOST=$HOST_IP ./scripts/cherami-setup-schema
```
Comment (venkat1109) [inline docker/entrypoint.sh]: where is this command used ?

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+FROM cassandra:3.9
+
+# get golang 1.7.4 (https://github.com/docker-library/golang/blob/18ee81a2ec649dd7b3d5126b24eef86bc9c86d80/1.7/Dockerfile)
+RUN apt-get update && apt-get install -y --no-install-recommends \
```
Comment (venkat1109) [inline docker/Dockerfile]: Would be cool to have a macos version as well.. may be later.

File: docker/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# start up cassandra
+pushd /
+./docker-entrypoint.sh cassandra
```
Comment (venkat1109) [inline docker/entrypoint.sh]: where is this file ?

File: docker/Dockerfile (hunk 3 above)
Comment (thuningxu) [inline docker/Dockerfile]: This works under MacOS. You need to install Docker for Mac.

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+VOLUME /var/lib/cherami-store
+EXPOSE 4922 4253 4240 4254 5425 6280 6189 6190 6191 6310
+
+COPY ./entrypoint.sh $CHERAMI_HOME/entrypoint.sh
+COPY ./docker_template.yaml $CHERAMI_HOME/config/docker_template.yaml
+RUN chmod a+x $CHERAMI_HOME/entrypoint.sh
+
+WORKDIR $CHERAMI_HOME
+ENTRYPOINT $CHERAMI_HOME/entrypoint.sh
+CMD ./cherami-server start all
```
Comment (thuningxu) [inline docker/Dockerfile]: I am still not quite clear about ENTRYPOINT vs CMD.

File: docker/entrypoint.sh (hunk 4 above)
Comment (thuningxu) [inline docker/entrypoint.sh]: This image inherits "cassandra:3.9" (`FROM cassandra:3.9`). `docker-entrypoint.sh` is from that image.

File: docker/entrypoint.sh (hunk 2 above)
Comment (thuningxu) [inline docker/entrypoint.sh]: which command?

File: docker/entrypoint.sh (hunk 1 above)
Comment (thuningxu) [inline docker/entrypoint.sh]: Not sure what's the right way to do that. Maybe redo ./scripts/cherami-setup-schema to accept parameters. In that case it should be in a different PR.

**Comment by thuningxu** [comment]
Regarding making RingHosts configurable: this particular image is for single instance, as it also includes Cassandra inside. For cluster deployment I would make a different image that contains Cherami only.

File: docker/Dockerfile (hunk 3 above)
Comment (venkat1109) [inline docker/Dockerfile]: ah, yes.. don't know what I was thinking before..

File: docker/entrypoint.sh (hunk 2 above)
Comment (venkat1109) [inline docker/entrypoint.sh]: Question was about CQLSH_HOST.. Forgot it was an env that cqlsh understands. nvm..

**Comment by thuningxu** [comment]
Added new image "thuningxu/cherami-server" that is suitable for cluster deployment. It does not contain Cassandra, and needs externally supplied config.
